src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+
+namespace Microsoft.FeatureManagement.Telemetry.ApplicationInsights
+{
+    /// <summary>
+    /// Extensions used to add feature management publisher functionality.
+    /// </summary>
+    public static class ServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Adds an event publisher that publishes feature evaluation events to Application Insights.
+        /// </summary>
+        /// <param name="services">The service collection that feature management services are added to.</param>
+        /// <returns>The <see cref="IServiceCollection"/> that was given as a parameter, with the publisher added.</returns>
+        public static IServiceCollection AddFeatureManagementTelemetryPublisherApplicationInsights(this IServiceCollection services)
```

Comment:
Quite a long name :) I am wondering if it makes sense to have something like
```cs
using Microsoft.FeatureManagement;

builder.Services.AddFeatureManagement()
    .AddFeatureFilter<PercentageFilter>()
    .AddFeatureManagementTelemetry<ApplicationInsightsTelemetryPublisher>();

```

Reply:
Another option... make telemetry more discoverable.

```cs
using Microsoft.FeatureManagement;

builder.Services.AddFeatureManagement<ApplicationInsightsTelemetryPublisher>()
                .AddFeatureFilter<PercentageFilter>();
```

Reply:
I like the `.AddFeatureManagementTelemetry<ApplicationInsightsTelemetryPublisher>();` approach. It's just as long, but it's in line with how filters are added and it would handle future publishers (custom or first party). I'll switch to that.

Reply:
I don't think it makes sense to add generics just for API naming purposes.

Reply:
I think it's accepting that it should be generic. Now developers can do:
```csharp
public class CustomPublisher : ITelemetryPublisher
```

```csharp
services.AddFeatureManagement()
    .AddFeatureManagementTelemetry<CustomPublisher>();
```

Rather than add it as a singleton.

Reply:
What's our publishing behavior on:
```cs
services.AddFeatureManagement()
    .AddFeatureManagementTelemetry<CustomPublisher>();
    .AddFeatureManagementTelemetry<CustomPublisher2>();
```cs

Reply:
My understanding is that telemetry publisher is option of telemetry, not a separate service. Similar to what we do with ConfigureKeyVault.

```cs
services.AddFeatureManagement(o=> 
{
   o,ConfigureTelemetry(...)
});
```

Reply:
Currently `.AddFeatureManagement()` returns an `IFeatureManagementFactory`. This factory is where `.AddFeatureFilter` lives.

Meaning this works:
```csharp
services
    .AddSingleton(...)
    .AddFeatureManagement()
    .AddFeatureFilter(...)
```

But this wouldn't
```csharp
services
    .AddFeatureManagement()
    .AddFeatureFilter()
    .AddSingleton()
```

If we do add the `.AddFeatureManagementTelemetry` it should probably be within that factory.

> What's our publishing behavior on:

We currently only take the single one DI gives. We should probably support multiple but haven't done so yet. 

>  telemetry is option of feature management

I think you're right, but I prefer parity with the existing FeatureFilters.

Reply:
> My understanding is that telemetry is option of feature management, not a separate service. Similar to what we do with ConfigureKeyVault.
> 
> ```cs
> services.AddFeatureManagement(o=> 
> {
>    o.ConfigureTelemetry(...)
> });
> ```

That's a pattern in app configuration .net provider. Not something that appears in this library. See how feature filter registration works in the example below.

```
services
    .AddFeatureManagement()
    .AddFeatureFilter<MyCustomFilter1>()
    .AddFeatureFilter<MyCustomFilter2>();
```

If we want to be symmetric with our telemetry publisher design then I would imagine a method like `AddTelemetryPublisher<MyCustomTelemetryPublisher>()`

However, since application insights is built-in, I don't think it should be a generic but rather a method like

`AddApplicationInsightsTelemetryPublisher`

So finally, I expect code like

```
services
    .AddFeatureManagement()
    .AddFeatureFilter<MyCustomFilter1>()
    .AddFeatureFilter<MyCustomFilter2>()
    .AddApplicationInsightsTelemetryPublisher();
```

Reply:
For built-in filters, we still do

```cs
services.AddFeatureManagement();
        .AddFeatureFilter<TargetingFilter>();
```

The benefit is that both first-party and third-party are handled in the same way, and third parties don't have to invent a new method to add their stuff.

Reply:
BTW, we are doing two things here
1. A telemetry extensibility
2. A telemetry extension for AppInsights

I expect a standard mechanism to add custom telemetry extensions is part of Job 1.

Reply:
Agree. I do think the following usage should be possible.

```
services
    .AddFeatureManagement()
    .AddFeatureFilter<MyCustomFilter1>()
    .AddFeatureFilter<MyCustomFilter2>()
    .AddTelemetryPublisher<ApplicationInsightsTelemetryPublisher>();
```

but the application insights package has the opportunity to make it even easier for the user by exposing what I suggested.


```
services
    .AddFeatureManagement()
    .AddFeatureFilter<MyCustomFilter1>()
    .AddFeatureFilter<MyCustomFilter2>()
    .AddApplicationInsightsTelemetryPublisher();
```

If we do like the syntax sugar I'm fine to just have `AddTelemetryPublisher<ApplicationInsightsTelemetryPublisher>()`

Reply:
I prefer 1P=3P approach.

Reply:
3rd party is free to publish their own package implementing a publisher and an extension method to make it discoverable. So both cases are 1p=3p

Reply:
Since we're aligned that `AddTelemetryPublisher<ApplicationInsightsTelemetryPublisher>()` should be possible, that's the only option I'll include in this PR. 

If we decide to add the other extension, we can do a separate PR.

Reply:
Fine with me.